storage/operation/cluster.go

```diff
@@ -29,12 +55,45 @@ func LookupClusterBlockHeight(r storage.Reader, clusterID flow.ChainID, height u
 	return RetrieveByKey(r, MakePrefix(codeFinalizedCluster, clusterID, height), blockID)
 }
 
-// UpsertClusterFinalizedHeight updates (overwrites!) the latest finalized cluster block height for the given cluster.
-func UpsertClusterFinalizedHeight(lctx lockctx.Proof, w storage.Writer, clusterID flow.ChainID, number uint64) error {
+func BootstrapClusterFinalizedHeight(lctx lockctx.Proof, rw storage.ReaderBatchWriter, clusterID flow.ChainID, number uint64) error {
+	if !lctx.HoldsLock(storage.LockInsertOrFinalizeClusterBlock) {
+		return fmt.Errorf("missing lock: %v", storage.LockInsertOrFinalizeClusterBlock)
+	}
+
+	key := MakePrefix(codeClusterHeight, clusterID)
+
+	var existing uint64
+	err := RetrieveByKey(rw.GlobalReader(), key, &existing)
+	if err == nil {
+		return fmt.Errorf("finalized height already initialized: %d", existing)
+	}
+
+	if !errors.Is(err, storage.ErrNotFound) {
```

Comment:
Es explained in [my comment above](https://github.com/onflow/flow-go/pull/7954#discussion_r2392567471), I would not include an `else` statement here.  

```suggestion
	// We do NOT want to continue with the WRITE UNLESS `storage.ErrNotFound` was received when checking for existing data.
	if !errors.Is(err, storage.ErrNotFound) {
```